File: session4/lecture102/beakjoon_1987.js
const input = require('fs').readFileSync(0, 'utf8').trim().split('\n');

const [R, C] = input[0].split(' ').map(Number);
const a = input.slice(1, 1 + R).map((line) => line.trim());

const visited = new Array(26).fill(0);
const dy = [-1, 0, 1, 0];
const dx = [0, 1, 0, -1];
let ret = 0;

// y, x : 현재 위치
// cnt  : 지금까지 지나온 칸 수
function go(y, x, cnt) {
  ret = Math.max(ret, cnt);
  for (let i = 0; i < 4; i++) {
    const ny = y + dy[i];
    const nx = x + dx[i];
    
    if (ny < 0 || ny >= R || nx < 0 || nx >= C) continue;
    const next = a[ny].charCodeAt(nx) - 65; // 'A'
    if (visited[next] === 0) {
      visited[next] = 1;
      go(ny, nx, cnt + 1);
      visited[next] = 0;
    }
  }
}

visited[a[0].charCodeAt(0) - 65] = 1;
go(0, 0, 1);
console.log(ret);

File: session5/lecture124/beakjoon_1987.cpp
#include <bits/stdc++.h>
using namespace std;
int R, C, ret, ny, nx, visited[30];
char a[21][21];
const int dy[] = {-1, 0, 1, 0};
const int dx[] = { 0, 1, 0, -1}; 

void go(int y, int x, int cnt){
    ret = max(ret, cnt);
    for(int i = 0; i < 4; i++){
      // lecture102 에서 풀었던 코드를 비트마스킹 기법으로 다시 작성해본다.
      // ny = y + dy[i], nx = x + dx[i];
      // if(ny < 0 || ny >= R || nx < 0 || nx >= C) continue;
      
      // int next = (int)(a[ny][nx] - 'A');
      // if(visited[next] == 0){
      //     visited[next] = 1; 
      //     go(ny, nx, cnt + 1);
      //     visited[next] = 0;  
      // } 

        ny = y + dy[i], nx = x + dx[i];
        if(ny < 0 || ny >= R || nx < 0 || nx >= C) continue;
        int _next = (1 << (int)(a[ny][nx] - 'A'));

        // num & _next 는 num과 _next가 겹치는 부분이 있는지 확인하는 것이다.
        // 겹치는 부분이 없다면 (num & _next) == 0 이
        // 겹치는 부분이 있다면 (num & _next) != 0 이 된다. 
        if((num & _next) == 0) {
          go(ny, nx, num | _next, cnt + 1);
        }
    }
    return;
}
int main(){
    cin >> R >> C;
    for(int i = 0; i < R; i++){
        for(int j = 0; j < C; j++){
            cin >> a[i][j];
        }
    }
    visited[(int)a[0][0] - 'A'] = 1;
    go(0, 0, 1);
    cout << ret << '\n';
    return 0;
}
